dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalSegment.cpp

```diff
@@ -163,71 +163,85 @@ SegmentPair DeltaMergeStore::segmentSplit(DMContext & dm_context, const SegmentP
     /// And only saved data in this segment will be filtered by the segment range in the merge process,
     /// unsaved data will be directly copied to the new segment.
     /// So we flush here to make sure that all potential data left by previous split operation is saved.
-    while (!left->flushCache(dm_context))
-    {
-        // keep flush until success if not abandoned
-        if (left->hasAbandoned())
-        {
-            LOG_FMT_INFO(log, "Merge - Give up segmentMerge because left abandoned, left={} right={}", left->simpleInfo(), right->simpleInfo());
-            return;
-        }
-    }
-    while (!right->flushCache(dm_context))
+    for (const auto & seg : ordered_segments)
     {
-        // keep flush until success if not abandoned
-        if (right->hasAbandoned())
+        size_t sleep_ms = 5;
+        while (true)
         {
-            LOG_FMT_DEBUG(log, "Merge - Give up segmentMerge because right abandoned, left={} right={}", left->simpleInfo(), right->simpleInfo());
-            return;
+            if (seg->hasAbandoned())
+            {
+                LOG_FMT_INFO(log, "Merge - Give up segmentMerge because segment abandoned, segment={}", seg->simpleInfo());
+                return {};
+            }
+
+            if (seg->flushCache(dm_context))
+                break;
+
+            // Else: retry. Flush could fail. Typical cases:
+            // #1. The segment is abandoned (due to an update is finished)
+            // #2. There is another flush in progress, for example, triggered in background
+            // Let's sleep 5ms ~ 100ms and then retry flush again.
+            std::this_thread::sleep_for(std::chrono::milliseconds(sleep_ms));
+            sleep_ms = std::min(sleep_ms * 2, 100);
         }
     }
 
-    SegmentSnapshotPtr left_snap;
-    SegmentSnapshotPtr right_snap;
+    std::vector<SegmentSnapshotPtr> ordered_snapshots;
+    ordered_snapshots.reserve(ordered_segments.size());
     ColumnDefinesPtr schema_snap;
 
     {
         std::shared_lock lock(read_write_mutex);
 
-        if (!isSegmentValid(lock, left))
+        for (const auto & seg : ordered_segments)
         {
-            LOG_FMT_DEBUG(log, "Merge - Give up segmentMerge because left not valid, left={} right={}", left->simpleInfo(), right->simpleInfo());
-            return;
+            if (!isSegmentValid(lock, seg))
+            {
+                LOG_FMT_DEBUG(log, "Merge - Give up segmentMerge because not valid, segment={}", seg->info());
+                return {};
+            }
         }
-        if (!isSegmentValid(lock, right))
+
+        for (const auto & seg : ordered_segments)
         {
-            LOG_FMT_DEBUG(log, "Merge - Give up segmentMerge because right not valid, left={} right={}", left->simpleInfo(), right->simpleInfo());
-            return;
-        }
+            // TODO: Should we ensure the ordering of "segments" first?
```

Comment:
The segments ordering is checked inside `Segment::prepareMerge`. Seems it is good enough?
https://github.com/pingcap/tiflash/blob/9a975379a502fead7e6c148de2597c0faef519dd/dbms/src/Storages/DeltaMerge/Segment.cpp#L1273-L1277